File: 0057_insert_interval/insert.py
from typing import List


def insert(intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
    """
    Time: O(n)
    Space: O(n)
    """
    res = []
    i, n = 0, len(intervals)
    start, end = newInterval
    while i < n and intervals[i][1] < start:
        res.append(intervals[i])
        i += 1
    while i < n and intervals[i][0] <= end:
        start = min(start, intervals[i][0])
        end = max(end, intervals[i][1])
        i += 1
    res.append([start, end])
    while i < n and intervals[i][0] > end:
        res.append(intervals[i])
        i += 1
    return res

File: 0057_insert_interval/insert.java
import java.util.List;
import java.util.ArrayList;

class Solution {
    // Time: O(n)
    // Space: O(n)
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> res = new ArrayList<>();
        int i = 0, n = intervals.length;
        int start = newInterval[0], end = newInterval[1];
        while (i < n && intervals[i][1] < start) {
            res.add(new int[] { intervals[i][0], intervals[i][1] });
            i++;
        }
        while (i < n && intervals[i][0] <= end) {
            start = Math.min(start, intervals[i][0]);
            end = Math.max(end, intervals[i][1]);
            i++;
        }
        res.add(new int[] { start, end });
        while (i < n && intervals[i][0] > end) {
            res.add(new int[] { intervals[i][0], intervals[i][1] });
            i++;
        }
        return res.toArray(new int[res.size()][2]);
    }
}